modules/internal/BytesStringCommon.chpl

```diff
@@ -719,27 +719,45 @@ module BytesStringCommon {
         const localNeedle = needle.localize();
         const needleLen = localNeedle.buffLen;
 
-        // i *is not* an index into anything, it is the order of the element
-        // of view we are searching from.
-        const numPossible = xLen - nLen + 1;
-        const searchSpace = if fromLeft
-            then 0..#(numPossible)
-            else 0..#(numPossible) by -1;
-        for i in searchSpace {
-          // j *is* the index into the localNeedle's buffer
-          for j in 0..#nLen {
-            const idx = view.orderToIndex(i+j); // 0s based idx
-            if x.buff[idx] != localNeedle.buff[j] then break;
-
-            if j == nLen-1 {
-              if count {
-                localRet += 1;
-              } else { // find
-                localRet = view.orderToIndex(i);
-              }
+        const numChar = 256;
+
+        // Preprocessing bad character heurestic
+        var badChar = c_calloc(c_int, numChar);
+
+        for i in 0..#numChar {
+          badChar[i] = nLen: int(32);
+        }
+
+        for i in 0..#nLen {
+          badChar[localNeedle.byte(i)] = (nLen - 1 - i):int(32);
+        }
+        // End preprocessing
+
+        var idx = 0;
+
+        while idx <= (xLen - nLen) {
+          var j = nLen - 1;
+
+          while(j >= 0) {
+            const idx1 = view.orderToIndex(idx+j);
```

Comment:
One (maybe too optimistic?) idea is to get rid of this index adjustments somehow. And maybe the solution is that we use this algorithm only if a search is done on the full string. This `view` logic is there to support searches on substrings in a string.

Reply:
@e-kayrakli I am not sure I fully understood your idea.
I think maybe this is about not using the algorithm for the strided search?

Reply:
Let's not worry too much about this right now, and first prove that this algorithm beats what we have with some large needles.

But to elaborate on the idea: (and I need to remind myself about some of the nuts-and-bolts here) we have this `view` and call to `orderToIndex` on it mostly to be able to do things like `"lorem ipsum".find("ip", region=2..7)` to search for `"ip"` only within a part of the string and not as a whole. I don't think this is a common case, and in the common case this call to `orderToIndex` adds extra (1) call (unless it is inlined) (2) some math costs. They are both innocent things if you have a long-running code, but may not be so here, especially since some of these calls appear in the deep loop nests (so we likely call them a lot).

As I said; I am only about 90 percent sure about some of the claims that I make here, and at this point this is more of a thing that we should keep in mind, rather than take immediate action on it.